### PR TITLE
Intoducing bfloat16 type

### PR DESCRIPTION
**Pull request by izdeby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21523 [WIP] Enabled BFloat16 storage
* **#21522 [WIP] Intoducing bfloat16 type**

**Introducing BFloat16 tensor type**
BFoat16 is a float32 truncated to its first 16 bits. So it has the same 8 bits for exponent, and only 7 bits for mantissa. It is therefore easy to convert from and to float32, and because it has basically the same range as float32, it minimizes the risks of having NaNs or exploding/vanishing gradients when switching from float32. In addition, you can use the bfloat16 format to accurately represent all integers [-256, 256], which means you can encode an int8 in bfloat16 without loss of accuracy.

In this PR: 
- Introduced BFloat16 type

Differential Revision: [D15819369](https://our.internmc.facebook.com/intern/diff/D15819369)

### CODE REVIEW

File: c10/util/BFloat16.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#include <math.h>
+#include <cmath>
+#include <stdint.h>
+#include <inttypes.h>
+#include <stdio.h>
+
+namespace c10 {
+
+namespace detail {
+  inline float f32_from_bf16(uint16_t src) {
```
Comment (gchanan) [inline c10/util/BFloat16.h]: this is strange because its `from_bf16`, but this doesn't take a BFloat16.  The half equivalent is `from_bits` which makes more sense.

File: c10/util/BFloat16.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+// Defines the bloat16 type (brain floating-point). This representation uses
+// 1 bit for the sign, 8 bits for the exponent and 7 bits for the mantissa.
+
+#include <c10/macros/Macros.h>
+#include <iostream>
```
Comment (gchanan) [inline c10/util/BFloat16.h]: which of these includes do you actually need?

File: c10/util/BFloat16.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+namespace detail {
+  inline float f32_from_bf16(uint16_t src) {
+    float res = 0;
+    uint16_t* tmp = reinterpret_cast<uint16_t*>(&res);
+    tmp[0] = 0;
+    tmp[1] = src;
+    return res;
+  }
+
+  inline uint16_t bf16_from_f32(float src) {
```
Comment (gchanan) [inline c10/util/BFloat16.h]: same naming thing here.

File: c10/util/BFloat16.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    uint16_t* res = reinterpret_cast<uint16_t*>(&src);
+    return res[1];
+  }
+} // namespace detail
+
+struct alignas(2) BFloat16 {
+  uint16_t val_;
+
+  BFloat16() = default;
```
Comment (gchanan) [inline c10/util/BFloat16.h]: can we do the same thing wrt HIP as we are doing in Half?

File: c10/util/BFloat16-inl.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+inline C10_HOST_DEVICE BFloat16::operator float() const {
+  return detail::f32_from_bf16(val_);
+}
+
+// CUDA intrinsics
+
+#if defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 350)
+inline __device__ BFloat16 __ldg(const BFloat16* ptr) {
+    return __ldg(reinterpret_cast<const BFloat16*>(ptr));
```
Comment (gchanan) [inline c10/util/BFloat16-inl.h]: is this used?  This doesn't look correct.

File: c10/util/BFloat16.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  inline C10_HOST_DEVICE BFloat16(float value);
```
Comment (gchanan) [inline c10/util/BFloat16.h]: can we make these explicit for now?  Would that make doing anything painful?

File: c10/util/BFloat16.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  inline float f32_from_bf16(uint16_t src) {
+    float res = 0;
+    uint16_t* tmp = reinterpret_cast<uint16_t*>(&res);
+    tmp[0] = 0;
+    tmp[1] = src;
+    return res;
+  }
+
+  inline uint16_t bf16_from_f32(float src) {
+    if (std::isnan(src)) {
```
Comment (gchanan) [inline c10/util/BFloat16.h]: if I'm understanding the format correctly, this shouldn't be necessary.  Can you show an example where it is?

Also, does your implementation handle denormals correctly?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: c10/util/BFloat16-inl.h (hunk 5 above)
Comment (izdeby) [reply c10/util/BFloat16-inl.h]: Weird. I was getting a lot of errors due to _ldg and explicit constructors but cant repro those anymore. Removing it until needed.

File: c10/util/BFloat16.h (hunk 6 above)
Comment (izdeby) [reply c10/util/BFloat16.h]: [This line](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THStorageCopy.cpp#L29) gives this error: error: invalid static_cast from type ‘c10::BFloat16’ to type ‘int32_t {aka int}’

File: c10/util/BFloat16-inl.h (hunk 5 above)
Comment (izdeby) [reply c10/util/BFloat16-inl.h]: It will be needed for [this](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCTensorCopy.cu#L12).

Comment (izdeby) [reply c10/util/BFloat16-inl.h]: discussed offline: we dont need it and will follow the same path as bool does for the __ldg

File: c10/util/BFloat16.h [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+#include <c10/macros/Macros.h>
+#include <cmath>
+
+namespace c10 {
+
+namespace detail {
+  inline float f32_from_bits(uint16_t src) {
+    float res = 0;
+    uint16_t* tmp = reinterpret_cast<uint16_t*>(&res);
```
Comment (ezyang) [inline c10/util/BFloat16.h]: Technically, doing this reinterpret cast violates the strict aliasing rule. The portable way to do this conversion is using a `memcpy`; there's some examples of this in `Half.h`

File: c10/test/util/bfloat16_test.cpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+      uint32_t fraction
+  ) {
+      uint32_t bytes;
+      bytes = 0;
+      bytes |= sign;
+      bytes <<= 8;
+      bytes |= exponent;
+      bytes <<= 23;
+      bytes |= fraction;
+      return *(float*)&bytes;
```
Comment (ezyang) [inline c10/test/util/bfloat16_test.cpp]: This also violates strict aliasing.

**Review by ezyang — APPROVED** [review]
Looks reasonable enough

**Review by gchanan — APPROVED** [review]
this looks pretty good!

It would be great if:
1) you could do use memcpy, as suggested by @ezyang.
2) you look into big endianness; we don't test it (which is fine for now, you can even assert in your bfloat16_test for now), but it should be only a few line change and you can just test it once.

**Comment by izdeby** [comment]
@ezyang, @gchanan  used memcpy and added endianness check.

File: c10/test/util/bfloat16_test.cpp [hunk 10]
```diff
@@ -14,7 +14,10 @@ namespace {
       bytes |= exponent;
       bytes <<= 23;
       bytes |= fraction;
-      return *(float*)&bytes;
+
+      float res;
+      std::memcpy(&res, &bytes, sizeof(float));
```
Comment (gchanan) [inline c10/test/util/bfloat16_test.cpp]: it's probably slightly better to do sizeof(res) here.

File: c10/util/BFloat16.h [hunk 11]
```diff
@@ -5,21 +5,37 @@
 
 #include <c10/macros/Macros.h>
 #include <cmath>
+#include <cstring>
 
 namespace c10 {
 
 namespace detail {
+  inline bool isSmallEndian() {
+    int num = 1;
```
Comment (gchanan) [inline c10/util/BFloat16.h]: you should test this statically.

File: c10/util/BFloat16.h [hunk 12]
```diff
@@ -5,21 +5,37 @@
...
+  }
+
   inline float f32_from_bits(uint16_t src) {
     float res = 0;
-    uint16_t* tmp = reinterpret_cast<uint16_t*>(&res);
-    tmp[0] = 0;
-    tmp[1] = src;
+    uint32_t tmp = src;
+    if (isSmallEndian()) {
+      tmp <<= 16;
```
Comment (gchanan) [inline c10/util/BFloat16.h]: did you test this?  This doesn't seem right -- the bitshift operators should do the 'right' thing given the endianness of the system.

File: c10/util/BFloat16.h [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+    memcpy(&res, &tmp, sizeof(tmp));
+    return res;
+  }
+
+  inline uint16_t bits_from_f32(float src) {
+    uint32_t res;
+    memcpy(&res, &src, sizeof(res));
+    if (isSmallEndian()) {
+        return res >>= 16;
```
Comment (gchanan) [inline c10/util/BFloat16.h]: nit: spacing.

Comment (gchanan) [inline c10/util/BFloat16.h]: I'm still not convinced this is correct for big endian; I don't see anything you are doing here that is different on big endian and small endian (the code looks correct for small endian).

File: c10/util/BFloat16.h [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    } else {
+      return res;
+    }
+  }
+} // namespace detail
+
+struct alignas(2) BFloat16 {
+  uint16_t val_;
+
+  // HIP wants __host__ __device__ tag, CUDA does not
```
Comment (colesbury) [inline c10/util/BFloat16.h]: Use C10_HIP_HOST_DEVICE

File: c10/util/BFloat16-inl.h [hunk 15]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+
+namespace c10 {
+
+/// Constructors
+
+inline C10_HOST_DEVICE BFloat16::BFloat16(float value) {
+  val_ = detail::bits_from_f32(value);
```
Comment (colesbury) [inline c10/util/BFloat16-inl.h]: These are marked `C10_HOST_DEVICE` but call host-only functions. That's not going to work when you include this in a CUDA file.

File: c10/util/BFloat16.h [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+namespace detail {
+  inline bool isSmallEndian() {
+    int num = 1;
+    return *(char *)&num == 1;
+  }
+
+  inline float f32_from_bits(uint16_t src) {
+    float res = 0;
+    uint32_t tmp = src;
+    if (isSmallEndian()) {
```
Comment (colesbury) [inline c10/util/BFloat16.h]: This is wrong on big-endian systems. Please remove all the little-endian checks.

The TensorFlow code needs the little-endian check because they are reinterpret_cast a float* to uint16* which depends on the byte-order. I have no idea whey they implemented it that way -- it seems unnecessarily complicated as well as violating strict-aliasing rules.

However, none of the operations here depend on byte order, except the memcpy assumes the ordering matches between uint32 and float.

https://github.com/tensorflow/tensorflow/blob/9590c4c32dd4346ea5c35673336f5912c6072bf2/tensorflow/core/framework/bfloat16.cc#L36

**Review by colesbury — CHANGES_REQUESTED** [review]

File: c10/util/BFloat16.h (hunk 16 above)
Comment (colesbury) [reply c10/util/BFloat16.h]: See https://gcc.godbolt.org/z/ARxFXh

Note that with powerpc64 clang (trunk) [big endian] z and y do not match.

File: c10/util/BFloat16-inl.h (hunk 15 above)
Comment (jgong5) [reply c10/util/BFloat16-inl.h]: Suggest to use round-nearest-to-even (RNE) as the default rounding mode rather than truncating. From our experiments, RNE is essential to guarantee convergence. Moreover, Tensorflow also uses RNE by default:

https://github.com/tensorflow/tensorflow/blob/a650bdc7e68285179da8abc310d5ffda268230a9/tensorflow/core/lib/bfloat16/bfloat16.h#L66
https://github.com/tensorflow/tensorflow/blob/a650bdc7e68285179da8abc310d5ffda268230a9/tensorflow/core/lib/bfloat16/bfloat16.h#L184

**Comment by jgong5** [comment]
Would this PR consider operator overloading like `+-*/` for `bfloat16` data type like what is being done for `half`?

File: c10/util/BFloat16-inl.h (hunk 15 above)
Comment (gchanan) [reply c10/util/BFloat16-inl.h]: @jgong5 we should just make this consistent with our other rounds.  I believe we are doing half-to-even as of https://github.com/pytorch/pytorch/pull/17443.

I think this can be implemented in another PR though.

**Comment by gchanan** [comment]
@jgong5 I don't think we are against it; our current implementation plan is just to implement conversion, fill, etc. operators on bfloat16, and it shouldn't be necessary for that.  But as soon as we have a use case, I don't see a reason not to add the C++ operators.

File: c10/util/BFloat16.h (hunk 16 above)
Comment (izdeby) [reply c10/util/BFloat16.h]: you are right. thanks for the godbolt link.
retested: https://gcc.godbolt.org/z/yEF0sM

**Comment by izdeby** [comment]
@jgong5, we will have overloads for bunch of operators in future PRs as soon as we will start adding support for math and related methods.

**Review by gchanan — APPROVED** [review]

File: c10/util/BFloat16.h (hunk 14 above)
Comment (izdeby) [reply c10/util/BFloat16.h]: it [fails](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-build/35434/console ) with an error: "unknown type name '__host__'"

Comment (colesbury) [reply c10/util/BFloat16.h]: Maybe I messed up something with that macro... don't worry about it then.

**Comment by izdeby** [comment]
@ezyang, @gchanan, @colesbury can you guys, please, take another look at this PR? 
Due to a few build issues i had to update this PR several times.

Important to note:
1. we cant use memcpy in f32_from_bits and bits_from_f32 as it will fail build in HIP.
2. making isSmallEndian() a static check isn't '[clean](https://stackoverflow.com/questions/1583791/constexpr-and-endianness)'.  Should i implement it? 
3. verified big/small endianess via unit tests for small endian and [godbolt](https://gcc.godbolt.org/z/Bt03pI) for a big one.

File: c10/util/BFloat16.h [hunk 17]
```diff
@@ -6,44 +6,51 @@
...
+    int num = 1;
+    return *(char *)&num == 1;
+  }
 
-  struct from_bits_t {};
-  static constexpr from_bits_t from_bits() {
-    return from_bits_t();
+  inline C10_HOST_DEVICE float f32_from_bits(uint16_t src) {
+    float res = 0;
+    uint16_t* tmp = reinterpret_cast<uint16_t*>(&res);
```
Comment (ezyang) [inline c10/util/BFloat16.h]: Could we have a comment here explaining why we are violating strict aliasing? :)

File: c10/util/BFloat16.h (hunk 11 above)
Comment (ezyang) [reply c10/util/BFloat16.h]: You link https://stackoverflow.com/questions/1583791/constexpr-and-endianness presumably because you wanted to implement this using `constexpr`. But I don't think we actually are requiring you to implement it with constexpr. All we are asking is that there isn't actually a conditional check in the generated machine code. I don't know if the current implementation you have here, it is managed to be compiled away; if it is, that would be sufficient for me (it's a good thing to put in a comment.)

**Comment by ezyang** [comment]
LGTM modulo comments.

File: c10/util/BFloat16.h [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace detail {
+  inline C10_HOST_DEVICE float f32_from_bits(uint16_t src) {
+    float* res = 0;
+    uint32_t tmp = src;
+    tmp <<= 16;
+
+    // We should be using memcpy in order to respect the strict aliasing rule
+    // but it fails in the HIP environment.
+    res = reinterpret_cast<float*>(&tmp);
```
Comment (gchanan) [inline c10/util/BFloat16.h]: I think it's better if you `ifdef` here and do the correct thing in non-HIP and the incorrect thing in HIP.  And then we can just get rid of the ifdef once HIP is fixed and we'll already have the correct code written.  As you have it now, you'll have to remember what you did when that time comes.

**Review by gchanan — CHANGES_REQUESTED** [review]

**Comment by ngimel** [comment]
It might make sense to do rne conversion rather than rtz (via @benbarsdell) https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/bfloat16/bfloat16.h#L178-L184

**Comment by gchanan** [comment]
@ngimel that's discussed above (I think the discussion is hidden though).

@izdeby did you file an issue for tracking the rounding here?

File: c10/util/BFloat16.h [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+namespace c10 {
+
+namespace detail {
+  inline C10_HOST_DEVICE float f32_from_bits(uint16_t src) {
+    float res = 0;
+    uint32_t tmp = src;
+    tmp <<= 16;
+
+#ifdef __HIP_PLATFORM_HCC__
+    float* tempRes = 0;
```
Comment (colesbury) [inline c10/util/BFloat16.h]: This is assigning 0 to a pointer type, which I guess is fine but looks kinda weird

File: c10/util/BFloat16.h [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+    // We should be using memcpy in order to respect the strict aliasing rule
+    // but it fails in the HIP environment.
+    tempRes = reinterpret_cast<float*>(&tmp);
+    res = *tempRes;
+#else
+    std::memcpy(&res, &tmp, sizeof(tmp));
```
Comment (colesbury) [inline c10/util/BFloat16.h]: If I remember correctly, the CUDA compiler doesn't optimize this memcpy well. I believe you use this in device code in your subsequent PR. You may want to check the CUDA performance and/or use the HIP codepath for CUDA as well.

File: c10/test/util/bfloat16_test.cpp [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+
+    EXPECT_FLOAT_EQ(in, out);
+  }
+
+  TEST(BFloat16Conversion, BiggestDenormal) {
+    float in =  float_from_bytes(0, 0, 0x7FFFFF); // The largest subnormal number
+    c10::BFloat16 a = c10::BFloat16(in);
+    float out = c10::detail::f32_from_bits(a.x);
+
+    EXPECT_FLOAT_EQ(1.1663108e-38, out);
```
Comment (gchanan) [inline c10/test/util/bfloat16_test.cpp]: where does this number come from?

File: c10/util/BFloat16.h (hunk 20 above)
Comment (ngimel) [reply c10/util/BFloat16.h]: I think it's fine now, in cuda 9.2 at least https://godbolt.org/z/024phL

File: c10/test/util/bfloat16_test.cpp (hunk 21 above)
Comment (gchanan) [reply c10/test/util/bfloat16_test.cpp]: I'd comment on what you are trying to test here.

**Review by gchanan — APPROVED** [review]

File: c10/test/util/bfloat16_test.cpp (hunk 21 above)
Comment (izdeby) [reply c10/test/util/bfloat16_test.cpp]: Killed this test as it not too representative now when i think about it.
